backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java

```diff
@@ -125,7 +127,19 @@ private static String addHqlConditions(String hql, Set<UUID> bundleIds, Set<UUID
         if (checkEndpointType || checkInvocationResult) {
             List<String> subQueryConditions = new ArrayList<>();
             if (checkEndpointType) {
-                subQueryConditions.add("nh.endpointType IN (:endpointTypes)");
+                List<String> subQueryEndpointTypes = new ArrayList<>();
+                Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
+                Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
+
+                if (basicTypes.size() > 0) {
+                    subQueryEndpointTypes.add("nh.compositeEndpointType.type IN (:basicEndpointTypes)");
+                }
+
+                if (compositeTypes.size() > 0) {
+                    subQueryEndpointTypes.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
+                }
```

Comment:
Did you check if this subquery works for basic types? `CompositeEndpointType#equals` compares both `type` and `subType`. If `type == null` then this query could still work. I'm not sure of Hibernate behavior in that case, this needs to be confirmed.

Reply:
> If `type == null` then this query could still work

I suppose you mean `sub_type`? `type` should never be `null`

I'll double check - I recall I checked this before and when the `sub_type` was `null` it wasn't working well. But I guess the query evolved a bit and is worth to check if that's still needed.

Reply:
Yes, I tried to use the composite directly (with nulls) and it didn't work.
Fails this test:  https://github.com/RedHatInsights/notifications-backend/blob/93159fc5a67bc6e06ab7523822fa08c5075b12ad/backend/src/test/java/com/redhat/cloud/notifications/routers/EventResourceTest.java#L282

```
Expected :1
Actual   :0
```

```
diff --git a/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java b/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
index 96be5149..05042701 100644
--- a/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
+++ b/backend/src/main/java/com/redhat/cloud/notifications/db/repositories/EventRepository.java
@@ -127,19 +127,7 @@ public class EventRepository {
         if (checkEndpointType || checkInvocationResult) {
             List<String> subQueryConditions = new ArrayList<>();
             if (checkEndpointType) {
-                List<String> subQueryEndpointTypes = new ArrayList<>();
-                Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
-                Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
-
-                if (basicTypes.size() > 0) {
-                    subQueryEndpointTypes.add("nh.compositeEndpointType.type IN (:basicEndpointTypes)");
-                }
-
-                if (compositeTypes.size() > 0) {
-                    subQueryEndpointTypes.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
-                }
-
-                subQueryConditions.add("(" + String.join(" OR ", subQueryEndpointTypes) + ")");
+                subQueryConditions.add("nh.compositeEndpointType IN (:compositeEndpointTypes)");
             }
             if (checkInvocationResult) {
                 subQueryConditions.add("nh.invocationResult IN (:invocationResults)");
@@ -169,16 +157,7 @@ public class EventRepository {
             query.setParameter("endDate", Timestamp.valueOf(endDate.atTime(LocalTime.MAX))); // at end of day
         }
         if (endpointTypes != null && !endpointTypes.isEmpty()) {
-            Set<EndpointType> basicTypes = endpointTypes.stream().filter(c -> c.getSubType() == null).map(CompositeEndpointType::getType).collect(Collectors.toSet());
-            Set<CompositeEndpointType> compositeTypes = endpointTypes.stream().filter(c -> c.getSubType() != null).collect(Collectors.toSet());
-
-            if (basicTypes.size() > 0) {
-                query.setParameter("basicEndpointTypes", basicTypes);
-            }
-
-            if (compositeTypes.size() > 0) {
-                query.setParameter("compositeEndpointTypes", compositeTypes);
-            }
+            query.setParameter("compositeEndpointTypes", endpointTypes);
         }
         if (invocationResults != null && !invocationResults.isEmpty()) {
             query.setParameter("invocationResults", invocationResults);

```

Reply:
Do you have the SQL generated by Hibernate when that test fails?

Reply:
Get Ids
```
select
        event0_.id as col_0_0_ 
    from
        event event0_ 
    where
        event0_.account_id=? 
        and (
            event0_.bundle_id in (
                ?
            )
        ) 
        and (
            event0_.application_id in (
                ?
            )
        ) 
        and (
            lower(event0_.event_type_display_name) like ?
        ) 
        and (
            event0_.created between ? and ?
        ) 
        and (
            exists (
                select
                    1 
                from
                    notification_history notificati1_ 
                where
                    notificati1_.event_id=event0_.id 
                    and (
                        (
                            notificati1_.endpoint_sub_type, notificati1_.endpoint_type
                        ) in (
                            (
                                ?, ?
                            )
                        )
                    ) 
                    and (
                        notificati1_.invocation_result in (
                            ?
                        )
                    )
            )
        ) 
    order by
        event0_.created desc limit ?
```

Get events
```
    select
        distinct event0_.id as id1_10_0_,
        historyent1_.id as id1_15_1_,
        event0_.account_id as account_2_10_0_,
        event0_.application_display_name as applicat3_10_0_,
        event0_.application_id as applicat4_10_0_,
        event0_.bundle_display_name as bundle_d5_10_0_,
        event0_.bundle_id as bundle_i6_10_0_,
        event0_.created as created7_10_0_,
        event0_.event_type_id as event_t10_10_0_,
        event0_.event_type_display_name as event_ty8_10_0_,
        event0_.payload as payload9_10_0_,
        historyent1_.created as created2_15_1_,
        historyent1_.endpoint_sub_type as endpoint3_15_1_,
        historyent1_.endpoint_type as endpoint4_15_1_,
        historyent1_.details as details5_15_1_,
        historyent1_.endpoint_id as endpoint8_15_1_,
        historyent1_.event_id as event_id9_15_1_,
        historyent1_.invocation_result as invocati6_15_1_,
        historyent1_.invocation_time as invocati7_15_1_,
        historyent1_.event_id as event_id9_15_0__,
        historyent1_.id as id1_15_0__ 
    from
        event event0_ 
    left outer join
        notification_history historyent1_ 
            on event0_.id=historyent1_.event_id 
    where
        event0_.id in (
            null
        ) 
    order by
        event0_.created desc

```

Count
```    
select
        count(*) as col_0_0_ 
    from
        event event0_ 
    where
        event0_.account_id=? 
        and (
            event0_.bundle_id in (
                ?
            )
        ) 
        and (
            event0_.application_id in (
                ?
            )
        ) 
        and (
            lower(event0_.event_type_display_name) like ?
        ) 
        and (
            event0_.created between ? and ?
        ) 
        and (
            exists (
                select
                    1 
                from
                    notification_history notificati1_ 
                where
                    notificati1_.event_id=event0_.id 
                    and (
                        (
                            notificati1_.endpoint_sub_type, notificati1_.endpoint_type
                        ) in (
                            (
                                ?, ?
                            )
                        )
                    ) 
                    and (
                        notificati1_.invocation_result in (
                            ?
                        )
                    )
            )
        )
```

Reply:
I see why it doesn't work: `something IN (null)` doesn't work with Postgres, so the query generated by Hibernate will never work. Writing a native query for that would be overkill, let's keep the current code.